core/finality-grandpa/src/lib.rs

```diff
@@ -644,41 +853,129 @@ impl<B, E, Block: BlockT<Hash=H256>, RA, PRA> BlockImport<Block>
 	{
 		use authorities::PendingChange;
 
+		// we don't want to finalize on `inner.import_block`
+		let justification = block.justification.take();
+		let number = block.header.number().clone();
+		let hash = block.post_header().hash();
+		let parent_hash = *block.header.parent_hash();
+		let digest = block.header.digest().clone();
+		let is_live = self.authority_set_oracle.is_live();
+
+		let import_result = self.inner.import_block(block, new_authorities)?;
+		if import_result != ImportResult::Queued {
+			return Ok(import_result);
+		}
+
 		let maybe_change = self.api.runtime_api().grandpa_pending_change(
-			&BlockId::hash(*block.header.parent_hash()),
-			&block.header.digest().clone(),
+			&BlockId::hash(parent_hash),
+			&digest,
 		)?;
 
-		// when we update the authorities, we need to hold the lock
-		// until the block is written to prevent a race if we need to restore
-		// the old authority set on error.
-		let just_in_case = maybe_change.map(|change| {
-			let hash = block.post_header().hash();
-			let number = block.header.number().clone();
-
+		if let Some(change) = maybe_change {
 			let mut authorities = self.authority_set.inner().write();
-			let old_set = authorities.clone();
 			authorities.add_pending_change(PendingChange {
 				next_authorities: change.next_authorities,
 				finalization_depth: change.delay,
 				canon_height: number,
 				canon_hash: hash,
 			});
 
-			block.auxiliary.push((AUTHORITY_SET_KEY.to_vec(), Some(authorities.encode())));
-			(old_set, authorities)
-		});
+			let encoded = authorities.encode();
+			self.inner.backend().insert_aux(&[(AUTHORITY_SET_KEY, &encoded[..])], &[])?;
+		};
 
-		let result = self.inner.import_block(block, new_authorities);
-		if let Err(ref e) = result {
-			if let Some((old_set, mut authorities)) = just_in_case {
-				debug!(target: "afg", "Restoring old set after block import error: {:?}", e);
-				*authorities = old_set;
-			}
+		let enacts_change = self.authority_set.inner().read().enacts_change(number, |canon_number| {
```

Comment:
Probably out of scope, but still: the change that is checked here (and applied later in `finalize_block` call) contains the same set of authorities as in the `new_authorities` set? Or those are/could be different sets?

Reply:
Also seems that if importing block with `justification` assumes `finalization` of the block (that's what @rphmeier asked above, iiuc), then imo we don't need `new_authorities` and authorities cache at all - we just need to refuse light-import of blocks which contain `DigestItem::as_authorities_change` and no justification. Could you, please, verify me here, @rphmeier ?

Reply:
@svyatonik yep, I believe that for GRANDPA we can indeed take that kind of approach. We may still want the authorities cache for other consensus algorithms.

Reply:
although this is a bit of future work, because it would require tracking the authorship validator changes in GRANDPA and saving justifications for those too upon finality.

this is a little bit harder than the GRANDPA set changes because we may have a commit for block N+K where a set change happened in block N; unlike GRANDPA-set changes where honest voters can never vote beyond N. so constructing the justification may become inefficient without a good way to prove ancestry far back in the chain.